contrib/platform/src/com/sun/jna/platform/win32/COM/util/Convert.java

```diff
@@ -78,43 +134,171 @@ public static VARIANT toVariant(Object value) {
+                            targetClass = Date.class;
+                            break;
+                        case VT_BSTR:
+                        case VT_BYREF | VT_BSTR:
+                            targetClass = String.class;
+                            break;
+                        case VT_UNKNOWN:
+                        case VT_BYREF | VT_UNKNOWN:
+                            targetClass = com.sun.jna.platform.win32.COM.IUnknown.class;
+                            break;
+                        case VT_DISPATCH:
+                        case VT_BYREF | VT_DISPATCH:
+                            targetClass = IDispatch.class;
+                            break;
+                        case VT_BYREF | VT_VARIANT:
+                            targetClass = Variant.class;
+                            break;
+                        case VT_BYREF:
+                            targetClass = PVOID.class;
+                            break;
+                        case VT_BYREF | VT_DECIMAL:
+                            targetClass = OaIdl.DECIMAL.class;
+                            break;
+                        case VT_RECORD:
+                        default:
+                            if ((varType & VT_ARRAY) > 0
+                                    || ((varType & VT_SAFEARRAY) > 0)) {
+                                targetClass = OaIdl.SAFEARRAY.class;
+                            }
+                    }
                 }
-                Object vobj = value.getValue();
-                if (vobj != null && concreteClassRequested && targetClass.isAssignableFrom(vobj.getClass())) {
-                        return vobj;
+                
+                Object result;
+                if(Byte.class.equals(targetClass) || byte.class.equals(targetClass)) {
+                    result = value.byteValue();
+                } else if (Short.class.equals(targetClass) || short.class.equals(targetClass)) {
+                    result = value.shortValue();
+                } else if (Character.class.equals(targetClass) || char.class.equals(targetClass)) {
+                    result = (char) value.intValue();
+                } else if (Integer.class.equals(targetClass) || int.class.equals(targetClass)) {
+                    result = value.intValue();
+                } else if (Long.class.equals(targetClass) || long.class.equals(targetClass) || IComEnum.class.isAssignableFrom(targetClass)) {
+                    result = value.longValue();
+                } else if (Float.class.equals(targetClass) || float.class.equals(targetClass)) {
+                    result = value.floatValue();
+                } else if (Double.class.equals(targetClass) || double.class.equals(targetClass)) {
+                    result = value.doubleValue();
+                } else if (Boolean.class.equals(targetClass) || boolean.class.equals(targetClass)) {
+                    result = value.booleanValue();

```

Comment:
The code for _byte_ asks if the class is the wrapper or the primitive type, but the rest for the code does not as this for _short, int, long, double, float_ and _char_. Is there a special reason for this ? Shouldn't the code be:
```java
if ((Byte.class == targetClass) || (byte.class == targetClass)) {
    ... handle byte...
} else if ((Short.class == targetClass) || (short.class == targetClass)) {
    ... handle short...
} else if ((Integer.class == targetClass) || (int.class == targetClass)) {
    ... handle int ...
} else ... etc...
}
```
BTW, for class equality you don't need to use _equals_ - references are adecuate.